nameserver/zone.go

```diff
@@ -585,30 +583,45 @@ func (zone *ZoneDb) AddRecord(ident string, name string, ip net.IP) (err error)
 	return
 }
 
-// Delete all records for an ident
-func (zone *ZoneDb) DeleteRecordsFor(ident string) (err error) {
-	zone.mx.Lock()
-	defer zone.mx.Unlock()
-	Debug.Printf("[zonedb] Deleting records for ident '%s'", ident)
-	zone.deleteIdent(ident)
-	return
-}
-
-// Delete an IP in a ident
-func (zone *ZoneDb) DeleteRecord(ident string, ip net.IP) (err error) {
+// Delete matching records (uninitialised values act as wildcards)
+func (zone *ZoneDb) DeleteRecords(ident string, name string, ip net.IP) int {
 	zone.mx.Lock()
 	defer zone.mx.Unlock()
 
-	Debug.Printf("[zonedb] Deleting record '%s'/[%s]", ident, ip)
-	if _, found := zone.idents[ident]; found {
-		err = zone.idents[ident].deleteIP(ipToIPv4(ip))
-		if zone.idents[ident].empty() {
-			zone.deleteIdent(ident)
+	count := 0
+
+	if name != "" {
+		name = dns.Fqdn(name)
+	}
+
+	for identK, identV := range zone.idents {
+		if ident == "" || ident == identK {
+			for nameK, nameV := range identV.names {
+				nameModified := false
+				if name == "" || name == nameK {
+					for ipv4K, ipv4V := range nameV.ipv4 {
+						if ip.Equal(net.IP{}) || ip.Equal(ipv4K.toNetIP()) {
+							ipv4V.notifyIPObservers()
+							delete(nameV.ipv4, ipv4K)
+							nameModified = true
+							count++
+						}
```

Comment:
I was wondering if we could use methods like `zone.go/nameSet.deleteName()/deleteIP()`, `zone.go/name.deleteIP()`, etc. For example, maybe you could move the `ip.Equal(net.IP{})` comparison out of the loop and just `if ip.Equal(net.IP{}) { identV.name.deleteName(name) } else { nameV.deleteIP(ipv4K) }`...